faunadb-java-dsl/src/main/java/com/faunadb/client/types/Encoder.java

```diff
@@ -0,0 +1,130 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.types.Properties.Property;
+import com.faunadb.client.types.Value.ArrayV;
+import com.faunadb.client.types.Value.ObjectV;
+
+import java.lang.reflect.Array;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Stack;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Lists.newArrayListWithCapacity;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB object to {@link Value} encoder.
+ */
+public final class Encoder {
+  private Encoder() {
+  }
+
+  private final static Stack<Object> stack = new Stack<>();
```

Comment:
I completely forgot that, so refactoring this I won't be able to keep `ObjectCodec` class because we cannot keep that class cached since it has to carry an instance of `Encoder` in order to encode the fields recursively. I'll have to keep two functions cached one for decoder and one for encoder, that one receiving two arguments, the encoder instance and the object to encode.